index.bs

```diff
@@ -598,13 +598,15 @@ interface SpeechSynthesisUtterance : EventTarget {
 interface SpeechSynthesisEvent : Event {
     readonly attribute SpeechSynthesisUtterance utterance;
     readonly attribute unsigned long charIndex;
+    readonly attribute unsigned long? charLength;
```

Comment:
why null instead of just zero?

Reply:
This is related to the "return undefined" bits in https://w3c.github.io/speech-api/#speechsynthesisevent-attributes which don't actually make sense, it's not possible to return undefined through Web IDL. Using `elapsedTime` zero, `name` empty string would be fine, and `charLength` zero too. It's only for `charIndex` that zero may be ambiguous with a real value.

However, with `charLength` added, it would be possible to distinguish this case by checking if `charLength` is zero. @sjdallst do you think that would be OK? The existing Chromium implementation just returns the default values, right?

Reply:
Argh, the above doesn’t sound amazing if it’s going to require special IDL treatment. Maybe we need a more in-depth spec review?

Reply:
It won't be hard to find other things that aren't quite right or not precise enough, but this "return undefined" problem is quite confined, it doesn't interact with other problems in the spec AFAICT.

Reply:
I think that the current state of things is a little ambiguous. If the user agent neglects charIndex completely, the client will see "undefined". If I get a 0 for charIndex, is this because the word or sentence is at the beginning of the text, or is it because the speech synthesizer does not support the charIndex? Returning null would remove some of this ambiguity, so a check for undefined or null would unambiguously tell the client whether this feature is supported, thus allowing for feature checking. Assuming the user agent follows the spec, this would also tell the client who is at fault (user agent or speech synthesis engine)

The current norm is the ambiguous option. I don't know if clients actually code for this ambiguous case, do platform checking, or hope for the best and let themselves break when features aren't supported, but following this norm would probably allow for the easiest and quickest implementation across browsers, and would let us resolve the undefined issue. However, issues might come later when clients have to keep track of what combination of browsers and platforms support these features.

Reply:
Are you planning to implement this in Chromium, and have you had a chance to check what the current implementation does yet?

Reply:
I have reviewed both the iterations and I am also inclined to setting charLength = 0 because it seems more consistent with how charIndex = 0 and name=emptyString defaults are being used. And since 0 is an organic valid value of an unsigned long, it will be more smooth to implement this in the browser (passthrough between layers) than a null value. Additionally charLength=0 is not a valid value and so will not be ambiguous.

Reply:
> If I get a 0 for charIndex, is this because the word or sentence is at the beginning of the text, or is it because the speech synthesizer does not support the charIndex?...  so a check for undefined or null would unambiguously tell the client whether this feature is supported

I'm not sure I understand: If the user agent doesn't support this feature, then why would the attribute be there? Wouldn't checking the prototype chain (I.e.,  `"charLength" in SpeechSynthesisEvent.prototype` - or variations on that) be the right way to do this? That would remove the ambiguity without needing to have a `length` with two types.

Reply:
The problem arises when the user agent (chromium, etc) supports the feature, but the speech synthesizer does not. For example, Windows, Mac, and, I think, Android have clear implementations for character length from speech events, but Linux (as far as I know) does not. Therefore, it makes sense to implement this feature in Chromium because it is supported by many speech synthesis platforms, but because the Linux version of chromium cannot get the value, it will be stuck returning a default value. 

In order to do the prototype chain checking, I think a web platform like chromium would have to keep different versions of the SpeechSynthesisEvent for different platforms depending on whether or not that platform's speech synthesizer supports the attribute. This sounds like it would add a lot of unecessary code for any browser implementer that wanted to do this.

I think that if we just look at charLength, it's not a bad idea to return the default value. Word and sentence events are only fired when a word or sentence is about to spoken, and there aren't any speakable words of 0 characters. 

Maybe charIndex is something that should be looked at in another PR?

Reply:
> Linux version of chromium cannot get the value, it will be stuck returning a default value.

What's preventing this getting fixed on Linux tho? It seems like we are working around an OS limitation that could be fixed at the OS level at any time in the future.

Reply:
(also noting that `null` adds further fingerprinting surface because it reveals something unique about the OS's capabilities - or lack of... not a huge thing, but it's a thing additional to consider)

Reply:
While I'm leaning towards not nullable because I suspect it aligns with existing implementations, note that if we did make it nullable and then all OS limitations were resolved, it would just mean that null is never returned and we can eventually make it non-nullable again.

Reply:
Ok it seems to me that we're leaning towards not nullable. I'll go ahead and make the change; along with edits to the descriptions of the other attributes.